Re-add urscript topic

potiuk opened the pull request:
Here is the promised pull request with re-adding URscript topic for executing custom UR Scripts.

gavanderhoorn commented:
Should 1fd2234 and cf17088 be part of the PR?

potiuk commented:
I can cherry-pick my commit and make it separate one if you think it is beter, no problem, but I based it on top of the  TCP/IP option changes 1fd2234 and I also figured that the stable configuration 
cf17088 makes sense and it's good to have it.

potiuk commented:
OK. I realised now that you @Zagitta added some more changes today I cherry-picked and rebased the commit on top of those.

potiuk commented:
@Zagitta @gavanderhoorn -> I added another, related commit to the RE_ADD_URSCRIPT branch. During my tests of my own custom RG2_GRIPPER scripts I found out that whenever you upload a wrong/non-compilable  script, the pipeline will be closed (not dropped) from the robot/ursim side and the driver is left in inconsistent state (the RT pipeline is closed and no further uploads are allowed). I added the argument (enabled by default) to shutdown the driver whenever pipeline is closed and modified the launches to respawn the driver if it shuts down. I think it's a "clean" solution for both development and "production" use, but I am happy to discuss and amend the change if you think we should do it differently.

Zagitta commented on src/ros_main.cpp:
```diff
@@ -109,6 +110,7 @@ int main(int argc, char **argv)
 
   ROSController *controller(nullptr);
   ActionServer *action_server(nullptr);
+  URScriptHandler *urscript_handler(nullptr);
```
This should probably not be a pointer, `controller` and `action_server` are pointers here because they're exclusive of each other.
(You also forgot to delete the pointer although that doesn't really matter 😄)

potiuk replied:
Sure. I thought I keep the same approach, but no problem to make it local variable :)

Zagitta commented on CMakeLists.txt:
```diff
@@ -145,15 +145,15 @@ endif()
 ## Your package locations should be listed before other locations
 # include_directories(include)
-include_directories(include 
+include_directories(include
```
You have lots of weird whitespace issues, could you please fix these to minimize further agony if someone has to diff changes?

Zagitta commented on launch/ur_common.launch:
```diff
@@ -2,7 +2,7 @@
   Universal robot common bringup.  Starts ur driver node and robot state
   publisher (translates joint positions to propper tfs).
-  
+
```
Whitespace issue

Zagitta commented on src/ros/trajectory_follower.cpp:
```diff
@@ -16,14 +16,14 @@ def driverProg():
 		exit_critical
 	end
-	
+
```
Whitespace issue

potiuk commented:
Great. Thanks for review!.

**Whitespace issues**

The whitespace changes are from my IDE settings (IntelliJ ultimate). I got it set to remove trailing whitespaces and addd EOL at the end of file, I can remove those no problem. I will do it soon. I am the guest here, so I need to adapt to your style/convention. For example with curly braces, I definitely prefer to open the braces in the same line as the if/while etc. but since I am contributing to your code, I have to adapt :D. Mixing functional + whitespace changes is usually not good. So I will change it back (I already reconfigured my IDE to not change whitespace).

**Topics vs. service vs. actions.** 

I do not think you need to change. We are using many ROS topics/services/actions in our setup (and had quite a number of discussions about those) - we also started from the article you mentioned. I believe service is not a good choice for urscript interface - it's either topic or action that you should consider. Usually services are used to return some value that is readily available. As a client, you need to get some result back, but the result should be returned without blocking. Service should return quickly and it should never block the client (an example is to get latest received image or current joint state). The URScript documentation is not very explicit about the interface (just mentions about sending the scirpt), but from the observed behaviour - when you send the UR script to the ur control PC you can at most know if it compiles or not, and you know it only because the connection will be closed shortly after you sent the program :(. There is no way to get the compile feedback without blocking (not even mentioning the execution feedback). There is no structured feedback nor confirmation after program is submitted (I'd love to have it BTW.). 

The best approach I see for feedback is to open the reverse connection from URScript and use it to communicate back + handling timeouts: similar to what trajectory follower is doing. I am in the middle of implementing it for our RG2_Gripper node (providing close/open actions). Action would be much more appropriate than service, because it can take quite some time to finish running a custom script in many cases. But I think having that at driver level is not needed. 

I think using topic is quite OK for the driver implementation level. If you need to wait for the script to execute and provide action-like behaviour, you can implement the TCP connection back and action interface on top of it. I am doing it for RG2_gripper open/close actions (but then it's totally outside of the driver and it's quite ok like that). 

Maybe you should simply provide "howto" - refer to an implemented custom showing how to implement an action using URScript topic interface. I already discussed it in our team and we will be happy to open-source our rg2_connector node as a good example on how you can use modern_driver's URScript topic  + TCP/IP communication back to implement the action interface (I am implementing it in python as it makes so much easier to test and run).

Let me know what you think about it.

potiuk commented:
OK. @Zagitta @gavanderhoorn - I removed all whitespace changes and changed the URScirpt handler to be local variable rather than pointer. Besides discussion of topic-> service (action?) change it's good to go (but even if yoy think about changing topic to service or action, it shoudl rather be a separate change/pull request).

potiuk commented:
FYI (regarding the Service/Action interface) @Zagitta @gavanderhoorn: if you are interested, I got already some good hands-on with the URScript/robot. I already implemented and I am testing right now a separate, generic RG2 gripper node that uses ROS Action/actionlib to open/close the gripper. It includes returnning grip status after it finishes, and providiing feedback while gripper is being opened/closed. I am using the ur_modern_driver's /URScript topic to upload the script. It seems to work perfectly well :). It's pure python (rospy) + URScript solution.

I did it using similar approach as the trajectory follower (reverse TCP/IP connection from URScript back to the node) but I am sending feedback from the URScript back to the node and have 2-way communication (that's what I am planning to do to improve the trajectory follower). I will open-source is shortly (providing that I will be able to open source parts of the RG2 URScript (I retrieved it from the UR PC). But I can share it with you before (later today most likely) if you would like to take a look.

Zagitta commented:
First of all thanks for your contribution @potiuk 🎉
Secondly I'm convinced you know more about the topic/service/action situation so I'll go with way you've coded it.
Third I really like your idea of adding names to the pipelines to make the logging easier to understand!
